<a href="https://colab.research.google.com/github/Junaidahmed1993/Junaidahmed1993/blob/circleci-project-setup/HuggingFacce.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**The Purpose of this codebook is to apply the hugging face library and the Roberta xlm Model for Sentiment Analysis**

In [ ]:
!pip install transformers
!pip install pytorch-lightning


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.5/7.5 MB 17.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 268.8/268.8 kB 31.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.8/7.8 MB 46.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 54.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 725.0/725.0 kB 11.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 761.3/761.3 kB 43.9 MB/s eta 0:00:00


In [3]:
!pip install transformers

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.5/7.5 MB 35.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 268.8/268.8 kB 31.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.8/7.8 MB 106.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 64.5 MB/s eta 0:00:00


In [3]:
from transformers import pipeline

In [ ]:
# import torch
# import torch.nn.functional as F

In [ ]:
classifier = pipeline("sentiment-analysis")

No model was supplied, defaulted to distilbert-base-uncased-finetuned-sst-2-english and revision af0f99b (https://huggingface.co/distilbert-base-uncased-finetuned-sst-2-english).
Using a pipeline without specifying a model name and revision in production is not recommended.


In [ ]:
text = classifier("Hey, listen; either you do the work or you leave the job. There is not other way")
print(text)

[{'label': 'NEGATIVE', 'score': 0.9911495447158813}]


In [ ]:
text2 = classifier(["I am really impressed by the work you've done. I think, you are now prepare for senior position.",
                     "but, i think given your bad track record and violent behavior we are uanbel to offer you this senior position"])
print(text2)

[{'label': 'POSITIVE', 'score': 0.9997714161872864}, {'label': 'NEGATIVE', 'score': 0.9967049956321716}]


**1. The Hugging Face Model**

The Above Code is simple use of the hugging face transformer library, We can use a model which hugging face offers and undertake sentiment analysis

**CODE**

In [4]:
from transformers import pipeline
from transformers import AutoTokenizer, AutoModelForSequenceClassification
import torch
import torch.nn.functional as F

In [5]:
# What kind of Hugging Face Model or which Hugging Face model, you want to call out
model_name = "distilbert-base-uncased-finetuned-sst-2-english"


In [6]:
# After, specifying the model, it is of utmost necessary that, we put our model into transformer library. So we call the model as;
model = AutoModelForSequenceClassification.from_pretrained(model_name) # So, here, we call Automodelsequenceclassification from pretrained and
                                                                       # And we actually gave out the model which we wanted to use here.


# Here we also load the tokenizer. The tokenizer plays an important role in NLP analysis
tokenizer = AutoTokenizer.from_pretrained(model_name)


In [7]:
classifier = pipeline("sentiment-analysis", model=model, tokenizer=tokenizer)
resultss = classifier (["We would like to inform you that, you have been positively evaluated by selection committe.",
                        "So, you are now able to join our team as a senior data scientist"])

for result in resultss:
  print(result)

{'label': 'POSITIVE', 'score': 0.9901959300041199}
{'label': 'POSITIVE', 'score': 0.999346911907196}


**2. Let's Do some Setps Manually Using Both Tokenizer and Models**

**A. Tokenizer Functions**

In [8]:
tokens = tokenizer.tokenize("We would like to inform you that, you have been positively evaluated by selection committe")
token_ids = tokenizer.convert_tokens_to_ids(tokens)

# This may be little bit difficult to execute from start but, it's quite a easy as we see, it yeild into both words (tokens)
#  And thier numerical representation

**A.i. Direct Tokenization**

In [9]:
input_ids = tokenizer("We would like to inform you that, you have been positively evaluated by selection committe")

# This is easiet way to tokenize the text. However, We need to undertsand the properly, how does it works

**Let's Print the Results of All these three varaibles**

So, A and A.i are both same things. Its matter of how we execute the tokenization.

In [10]:
print (f' Token: {tokens}')
print (f'Token IDS {token_ids}')
print (f'Inputs {input_ids}')

 Token: ['we', 'would', 'like', 'to', 'inform', 'you', 'that', ',', 'you', 'have', 'been', 'positively', 'evaluated', 'by', 'selection', 'commit', '##te']
Token IDS [2057, 2052, 2066, 2000, 12367, 2017, 2008, 1010, 2017, 2031, 2042, 13567, 16330, 2011, 4989, 10797, 2618]
Inputs {'input_ids': [101, 2057, 2052, 2066, 2000, 12367, 2017, 2008, 1010, 2017, 2031, 2042, 13567, 16330, 2011, 4989, 10797, 2618, 102], 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]}


So, From first code, we have words as tokens and thier token ids.
In the Second code, we only have tokens.

The input_ids= 101 is begnning of strings and 102 and is the ending of strings

**LET's CALL THE MODEL**

In [13]:
x_train = "We would like to inform you that, you have been positively evaluated by selection committe"
batch = tokenizer(x_train, padding=True, truncation=True, max_length=512, return_tensors="pt")
print(batch)

{'input_ids': tensor([[  101,  2057,  2052,  2066,  2000, 12367,  2017,  2008,  1010,  2017,
          2031,  2042, 13567, 16330,  2011,  4989, 10797,  2618,   102]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]])}


**LET'S USE THE MODEL TO CALL TENSOR AND ANALZE SENTIMENTS**

In [16]:
with torch.no_grad():
  outputs = model(**batch)
  print(outputs)
  predictions= F.softmax(outputs.logits, dim=1)
  print(predictions)
  labels = torch.argmax(predictions, dim=1)
  print(labels)
  labels = [model.config.id2label[label_id] for label_id in labels.tolist()]
  print(labels)

SequenceClassifierOutput(loss=None, logits=tensor([[-2.3240,  2.4318]]), hidden_states=None, attentions=None)
tensor([[0.0085, 0.9915]])
tensor([1])
['POSITIVE']


In [ ]:
nu

**So, we have two methods**

**Pipeline**: The pipeline allow us to calcualte the resuls using the using two lines of code

**Maual:** The manual code which may be of 5-6 line would also help us to calcuklate the results.

**let's Save the results**

In [ ]:
save_directory = 'saved'
tokenizer.save_pretrained(save_directory)
model.save_pretrained(save_directory)
tokenizer = AutoTokenizer.from_pretrained(save_directory)
model = AutoModelForSequenceClassification.from_pretrained(save_directory)

In [1]:
print("hello worlf")

hello worlf


**let's Analyze Sentiment on my dataSet**

In [2]:
from transformers import pipeline
classifier = pipeline ("sentiment-analysis")

ModuleNotFoundError: ignored

In [1]:
text = classifier ("You are great person, I admire you")
text1= classifier ("you are evil persom, I despise you")
print(text)
print(text1)

NameError: ignored